src/main/kotlin/com/shankhadeepghoshal/kotlinfinaltask/service/Service.kt

```diff
@@ -0,0 +1,92 @@
+import com.shankhadeepghoshal.kotlinfinaltask.exception.DogNotFoundException
+import com.shankhadeepghoshal.kotlinfinaltask.pojo.Dog
+import com.shankhadeepghoshal.kotlinfinaltask.pojo.DogApiResponseGetImages
+import com.shankhadeepghoshal.kotlinfinaltask.pojo.Url
+import com.shankhadeepghoshal.kotlinfinaltask.repository.DogRepository
+import com.shankhadeepghoshal.kotlinfinaltask.repository.GET_ALL_DOG_BREED_URL
+import com.shankhadeepghoshal.kotlinfinaltask.repository.IMAGE_URL_BY_BREED_MAIN_NAME
+import com.shankhadeepghoshal.kotlinfinaltask.repository.UrlRepository
+import kotlinx.coroutines.flow.*
+import mu.KotlinLogging
+import org.springframework.http.MediaType
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+import org.springframework.web.bind.annotation.PathVariable
+import org.springframework.web.reactive.function.client.WebClient
+import org.springframework.web.reactive.function.client.awaitBody
+
+interface DogsApiService {
+    fun findAllDogs(): Flow<Dog>
+    suspend fun findDogsUrl(dogName: String): List<Url>?
+    suspend fun findAllDogsAndPersist()
+}
+
+@Service
+class DogsApiServiceImpl(
+    private val dogRepo: DogRepository,
+    private val urlRepository: UrlRepository,
+    private val restService: RestService,
+    private val objectMapper: ObjectMapper
+) : DogsApiService {
+    private val log = KotlinLogging.logger {}
+    override fun findAllDogs() = dogRepo.findAll()
+
+    @Throws(DogNotFoundException::class)
+    @Transactional
+    override suspend fun findDogsUrl(dogName: String): List<Url>? {
+        var urlsByDogNameFromDb = urlRepository.findUrlsGivenDogName(dogName)
+
+        if (urlsByDogNameFromDb.isNullOrEmpty()) {
+            val urlsByDogNameFromApi = restService.getUrlsDogName(dogName).message
+            if (urlsByDogNameFromApi.isNullOrEmpty()) {
+                log.error { "Url not found for $dogName" }
+                throw DogNotFoundException("Doge with given name not found")
+            } else {
+                val dogId = dogRepo.findIdsGivenBreedMainName(dogName)
+                urlsByDogNameFromApi.forEach { urlRepository.insertUrlForDogId(dogId, it) }
+                urlsByDogNameFromDb = urlRepository.findUrlsGivenDogName(dogName)
+                log.info { "Id found for $dogName is $dogId" }
+            }
+        }
+
+        return urlsByDogNameFromDb
+    }
+
+    @Transactional
+    override suspend fun findAllDogsAndPersist() {
+        if (dogRepo.findById(1) == null) {
+            objectMapper.readTree(restService.getAllDogs())
+                .get("message")
+                .fields()
+                .asSequence()
+                .map { entry ->
+                    Dog(name = entry.key)
+                }.forEach { dogRepo.save(it) }
+        }
+    }
+}
+
+@Service
+class RestService {
+    suspend fun getUrlsDogName(@PathVariable("breedName") breedName: String) =
```

Comment:
@PathVariable is redundant here and must be removed.

Reply:
Getting warning from IDE saying that path-variable parameter not consumed when the annotation is removed

Reply:
@PathVariable annotation does not do anything here, may be you have an ide cache issue? For comparison, there is no such an issue in my ide (as expected):
<img width="717" alt="Screenshot 2022-11-28 at 20 13 08" src="https://user-images.githubusercontent.com/20745324/204329118-7e4c59bf-5c97-42bc-af05-db31aebb081d.png">